qiskit/quantum_info/operators/operator.py

```diff
@@ -420,9 +423,16 @@ def from_circuit(
         # Convert circuit to an instruction
         instruction = circuit.to_instruction()
         op._append_instruction(instruction, qargs=qargs)
-        # If final layout is set permute output indices based on layout
+        # If final layout is set permute output indices based on that layout combined with initial layout
         if final_layout is not None:
-            perm_pattern = [final_layout._v2p[v] for v in circuit.qubits]
+            if layout is None:
+                perm_pattern = [final_layout._v2p[v] for v in circuit.qubits]
+            else:
+                perm_pattern = [
+                    layout.input_qubit_mapping[layout.initial_layout._p2v[final_layout._v2p[v]]]
+                    for v in circuit.qubits
+                ]
+                perm_pattern = [perm_pattern[i] for i in layout.initial_layout.get_physical_bits()]
```

Comment:
I'm wondering if we can use `final_index_layout()` (https://docs.quantum.ibm.com/api/qiskit/qiskit.transpiler.TranspileLayout#final_index_layout) here to simplify this?

Reply:
None of the `*_*_layout()` functions in `TranspileLayout` appear to yield the correct permutation pattern... :-/

Reply:
I think it would be the inverse of `final_index_layout`, because `final_index_layout` returns the final permutation of each qubit but we need the inverse of that to reverse the permutation for the operator construction. Does something like this work?:

```python
final_index_layout = layout.final_index_layout()
perm_pattern = [None]*len(final_index_layout)
for index, pos = enumerate(final_index_layout):
    perm_pattern[pos] = index
```

Reply:
Does not appear to be the case... :-/ One explicit example is
1. `[5, 1, 0, 2, 4, 3]`  - correct
2. `[5, 3, 0, 4, 1, 2]` - inverse `final_index_layout`
3. `[2, 4, 5, 1, 3, 0]` - `final_index_layout`
4. `[5, 4, 1, 0, 3, 2]` - `initial_index_layout`
My impression was that this code might become obsolete with https://github.com/Qiskit/qiskit/pull/11399 which is why I did not spend a terrible amount of time on it. :-) 

Do you want me to formulate this fix as a function of `final_index_layout` and `initial_index_layout`?

Reply:
I'm pretty sure the index layout functions are correct in their current form. Or at least at the time we added them we did a fair amount of testing around that. But I have zero confidence in any statements about handling permutations, so either my assertion that the inverse of `final_index_layout` is what we want here is incorrect, or `final_index_layout()` is not correctly written. I'm more inclined the believe the former, but I could see either being possible.

Reply:
Neither `initial_index_layout` nor `final_index_layout` are permutations nor are their inverses.  They're both layouts - it's correct and expected that neither alone is a suitable permutation within the same space here, because both are $V \leftrightarrow P$ bi-maps and not $P \to P$ nor $V \to V$ maps.

When we're comparing the untranspiled `qc` to the result of a transpilation `out`, you can map `out` back to the exact same indices as `qc` by using the `initial_index_layout` as an inverse permutation at the start, and `final_index_layout` as a forwards permutation at the end:
```python
from qiskit import QuantumCircuit
from qiskit.circuit.library import PermutationGate

def undo(transpiled):
    initial = transpiled.layout.initial_index_layout()
    final = transpiled.layout.final_index_layout()

    out = QuantumCircuit(transpiled.num_qubits)
    out.append(PermutationGate(initial).inverse(), out.qubits)
    out.compose(transpiled, out.qubits, inplace=True)
    out.append(PermutationGate(final), out.qubits)
    return out
```

Now:

```python
import itertools
import math
import numpy.random

from qiskit import QuantumCircuit, transpile
from qiskit.transpiler import CouplingMap
from qiskit.quantum_info import Operator

def rand_circ(qubits, gates):
    rng = numpy.random.default_rng()
    links = list(itertools.permutations(range(qubits), 2))

    qc = QuantumCircuit(qubits)
    for _ in range(gates):
        qc.rzx(rng.random()*2*math.pi, *rng.choice(links))
    return qc

qc = rand_circ(8, 100)
transpiled = transpile(qc, coupling_map=CouplingMap.from_line(qc.num_qubits))

assert Operator(qc) == Operator(undo(transpiled))
```

(note that neither `qc` nor `undo(transpiled)` have a set `layout`). I've just used the #10835 methods on `TranspileLayout` (which already included all the information to make all the correct permutations) and our existing `PermutationGate` conventions, rather than the #11399 methods on `Layout` that need extra information and have a different set of permutations.

I haven't done the ancilla suppression, but again, all the information is in `TranspileLayout` already.

Reply:
Just to be clear: I'm not necessarily suggesting building the new circuit the way I did in order to power `Operator.from_circuit` - if nothing else, I don't think that'd work too well with ancilla suppression.  I was just using my `undo` as an example to show how the #10835 `TranspileLayout` API integrates with the existing `PermutationGate` and the existing `Operator` matrix expansions.